File: services/crown-adapter/src/hardware.ts
import type { CrownConfig } from "./config.ts";
import { deviceStatusEvent, heartbeatEvent, makeEvent, type UnnormalizedEvent } from "./events.ts";
import type { CrownAccel, CrownClient, CrownEpoch } from "./neurosityClient.ts";
import type { Publisher } from "./publisher.ts";
import { CROWN_CHANNELS, sanitizeSamples, scoreQuality } from "./quality.ts";
import { deviceMsToNs } from "./timestamps.ts";

const MOTION_AXES = ["x", "y", "z"];
const DEVICE_ALIAS = "crown";

export type HardwareDeps = {
  client: CrownClient;
  publisher: Publisher;
  config: CrownConfig;
  stopped: () => boolean;
  durationMs: number;
  now?: () => number;
  sleep?: (ms: number) => Promise<void>;
};

type AccelState = {
  x: number;
  y: number;
  z: number;
  magnitude: number;
  hasSample: boolean;
};

export async function runCrownHardware(deps: HardwareDeps): Promise<void> {
  const now = deps.now ?? Date.now;
  const sleep = deps.sleep ?? defaultSleep;
  const started = now();
  const deadline = deps.durationMs > 0 ? started + deps.durationMs : 0;
  let sequence = 0;
  let chunksEmitted = 0;
  let packetLossCount = 0;
  let lastStartTime = Number.NEGATIVE_INFINITY;
  let lastDataMs = now();
  let lastHeartbeat = 0;
  let attempt = 0;
  let clockOffsetNs: number | null = null;
  const accel: AccelState = { x: 0, y: 0, z: 0, magnitude: 0, hasSample: false };

  const nextSeq = () => {
    const value = sequence;
    sequence += 1;
    return value;
  };

  const send = async (event: UnnormalizedEvent) => {
    await deps.publisher.send(event);
  };

  const maybeHeartbeat = async (status: string) => {
    const t = now();
    if (t - lastHeartbeat < deps.config.heartbeatSeconds * 1000) {
      return;
    }
    const age = t - lastDataMs;
    await send(
      heartbeatEvent(
        nextSeq(),
        (t - started) / 1000,
        Number.isFinite(age) ? age : null,
        deps.publisher.droppedCount(),
        status,
      ),
    );
    lastHeartbeat = t;
  };

  const timedOut = () => deadline > 0 && now() >= deadline;
  const halt = () => deps.stopped() || timedOut();

  try {
    while (!halt()) {
      try {
        await deps.client.login();
        attempt = 0;
        await send(deviceStatusEvent(nextSeq(), "healthy", "crown stream started", {}, DEVICE_ALIAS));
        await consumeLiveStreams({
          client: deps.client,
          config: deps.config,
          accel,
          halt,
          send,
          nextSeq,
          maybeHeartbeat,
          onEpoch: async (epoch) => {
            const startTime = epoch.info?.startTime;
            if (typeof startTime === "number" && Number.isFinite(startTime) && startTime <= lastStartTime) {
              return;
            }
            if (typeof startTime === "number" && Number.isFinite(startTime)) {
              lastStartTime = startTime;
            }
            const events = convertEpoch(epoch, accel, {
              sequence: nextSeq,
              packetLossCount,
              chunksEmitted,
              motionArtifactThreshold: deps.config.motionArtifactThreshold,
              expectedChannels: deps.config.channels,
              expectedSamples: deps.config.samplesPerChunk,
              sampleRateHz: deps.config.sampleRateHz,
              shadowOnly: deps.config.shadowOnly,
              clockOffsetNs,
            });
            if (events.rejected) {
              packetLossCount += 1;
              return;
            }
            if (events.clockOffsetNs !== null) {
              clockOffsetNs = events.clockOffsetNs;
            }
            for (const event of events.events) {
              await send(event);
            }
            chunksEmitted += 1;
            lastDataMs = now();
          },
        });
        if (halt()) {
          break;
        }
        await send(
          deviceStatusEvent(nextSeq(), "degraded", "crown stream ended; reconnecting", {}, DEVICE_ALIAS),
        );
      } catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        if (isMissingCredentials(message)) {
          await send(deviceStatusEvent(nextSeq(), "offline", "missing Neurosity credentials", {}, DEVICE_ALIAS));
          throw error;
        }
        await send(
          deviceStatusEvent(
            nextSeq(),
            "degraded",
            "crown disconnected; reconnecting",
            { reconnect_attempt: attempt + 1 },
            DEVICE_ALIAS,
          ),
        );
      }

      if (halt()) {
        break;
      }
      const waitMs = backoffMs(attempt, deps.config.reconnectMaxSeconds);
      attempt += 1;
      clockOffsetNs = null;
      try {
        await deps.client.disconnect();
      } catch {
        // ignore disconnect errors during backoff
      }
      const waitEnd = now() + waitMs;
      while (!halt() && now() < waitEnd) {
        await maybeHeartbeat("degraded");
        await sleep(Math.min(100, waitEnd - now()));
      }
    }
  } finally {
    await send(deviceStatusEvent(nextSeq(), "offline", "adapter stopping", {}, DEVICE_ALIAS));
    try {
      await deps.client.disconnect();
    } catch {
      // ignore
    }
  }
}

export function backoffMs(attempt: number, capSeconds: number): number {
  const capMs = Math.max(1000, capSeconds * 1000);
  return Math.min(capMs, 1000 * 2 ** Math.max(0, attempt));
}

export function isMissingCredentials(message: string): boolean {
  return (
    message.includes("NEUROSITY_EMAIL") ||
    message.includes("NEUROSITY_PASSWORD") ||
    message.includes("NEUROSITY_DEVICE_ID")
  );
}

export function convertEpoch(
  epoch: CrownEpoch,
  accel: AccelState,
  opts: {
    sequence: () => number;
    packetLossCount: number;
    chunksEmitted: number;
    motionArtifactThreshold: number;
    expectedChannels: number;
    expectedSamples: number;
    sampleRateHz: number;
    shadowOnly: boolean;
    clockOffsetNs: number | null;
  },
): { events: UnnormalizedEvent[]; rejected: boolean; clockOffsetNs: number | null } {
  const samples = toChannelMajor(epoch.data, opts.expectedChannels);
  const channelNames =
    epoch.info?.channelNames?.length === opts.expectedChannels
      ? epoch.info.channelNames
      : [...CROWN_CHANNELS];
  const sampleCount = samples[0]?.length ?? 0;
  if (samples.length !== opts.expectedChannels || sampleCount === 0) {
    return { events: [], rejected: true, clockOffsetNs: opts.clockOffsetNs };
  }

  const { samples: cleaned, invalidChannels } = sanitizeSamples(samples);
  const startMs = epoch.info?.startTime;
  let sourceTimeNs: number | null = null;
  let clockOffsetNs = opts.clockOffsetNs;
  if (typeof startMs === "number" && Number.isFinite(startMs)) {
    sourceTimeNs = deviceMsToNs(startMs);
    if (clockOffsetNs === null) {
      clockOffsetNs = Date.now() * 1_000_000 - sourceTimeNs;
    }
  }

  const motionMagnitude = accel.hasSample ? accel.magnitude : 0;
  const quality = scoreQuality({
    samples: cleaned,
    packetLossCount: opts.packetLossCount,
    chunksSinceLoss: opts.chunksEmitted,
    motionMagnitude,
    motionArtifactThreshold: opts.motionArtifactThreshold,
  });
  if (invalidChannels.length > 0) {
    quality.flags.push("sanitized_non_finite");
  }
  if (sampleCount !== opts.expectedSamples) {
    quality.flags.push("unexpected_epoch_length");
  }

  const sampleRate = epoch.info?.samplingRate || opts.sampleRateHz;
  const events: UnnormalizedEvent[] = [];
  events.push(
    makeEvent({
      eventType: "biosignal.chunk",
      modality: "eeg",
      sequence: opts.sequence(),
      sourceTimeNs,
      quality: quality.score,
      payload: {
        sample_rate_hz: sampleRate,
        channel_names: channelNames,
        sample_count: sampleCount,
        samples: cleaned,
        units: "microvolts",
        filters_applied: ["crown_raw_filtered"],
        packet_loss_count: opts.packetLossCount,
        clock_confidence: sourceTimeNs === null ? 0.2 : 1,
        estimated_first_sample_ns: sourceTimeNs,
      },
    }),
  );

  if (accel.hasSample) {
    events.push(
      makeEvent({
        eventType: "motion.chunk",
        modality: "imu",
        sequence: opts.sequence(),
        sourceTimeNs,
        quality: quality.score,
        payload: {
          sample_rate_hz: sampleRate / Math.max(sampleCount, 1),
          axes: MOTION_AXES,
          sample_count: 1,
          samples: [[accel.x], [accel.y], [accel.z]],
          units: "g",
          magnitude: accel.magnitude,
        },
      }),
    );
  }

  events.push(
    makeEvent({
      eventType: "data.quality",
      modality: "eeg",
      sequence: opts.sequence(),
      sourceTimeNs,
      quality: quality.score,
      payload: {
        score: quality.score,
        components: {
          packet_quality: quality.packet_quality,
          channel_validity: quality.channel_validity,
          motion_penalty: quality.motion_penalty,
        },
        flags: quality.flags,
      },
    }),
  );

  if (opts.chunksEmitted > 0 && opts.chunksEmitted % 16 === 0 && sourceTimeNs !== null) {
    const artifact = motionMagnitude > opts.motionArtifactThreshold;
    events.push(
      makeEvent({
        eventType: "modality.feature",
        modality: "eeg",
        sequence: opts.sequence(),
        sourceTimeNs,
        quality: quality.score,
        payload: {
          feature_name: "eeg_shadow",
          window_start_ns: sourceTimeNs,
          window_end_ns: sourceTimeNs + deviceMsToNs(62),
          label: artifact ? "artifact" : "ok",
          confidence: quality.score,
          candidate_scores: {
            ok: artifact ? Math.max(0, 1 - quality.score) : quality.score,
            artifact: artifact ? quality.score : Math.max(0, 1 - quality.score),
          },
          model_id: "crown-shadow-v0",
          shadow_only: opts.shadowOnly,
        },
      }),
    );
  }

  return { events, rejected: false, clockOffsetNs };
}

export function toChannelMajor(data: number[][], expectedChannels: number): number[][] {
  if (!Array.isArray(data) || data.length === 0) {
    return [];
  }
  if (data.length === expectedChannels) {
    return data;
  }
  if (data[0]?.length === expectedChannels) {
    return Array.from({ length: expectedChannels }, (_, ch) => data.map((row) => Number(row[ch])));
  }
  return data;
}

function updateAccel(target: AccelState, sample: CrownAccel): void {
  const x = Number(sample.x);
  const y = Number(sample.y);
  const z = Number(sample.z);
  target.x = x;
  target.y = y;
  target.z = z;
  target.magnitude = Math.sqrt(x * x + y * y + z * z);
  target.hasSample = Number.isFinite(target.magnitude);
}

async function consumeLiveStreams(opts: {
  client: CrownClient;
  config: CrownConfig;
  accel: AccelState;
  halt: () => boolean;
  send: (event: UnnormalizedEvent) => Promise<void>;
  nextSeq: () => number;
  maybeHeartbeat: (status: string) => Promise<void>;
  onEpoch: (epoch: CrownEpoch) => Promise<void>;
}): Promise<void> {
  let active = true;
  const accelTask = drainIterable(opts.client.accelerometer(), opts.halt, () => active, (sample) => {
    updateAccel(opts.accel, normalizeAccel(sample));
  });
  const eegTask = drainIterable(opts.client.rawEpochs(), opts.halt, () => active, async (epoch) => {
    await opts.onEpoch(normalizeEpoch(epoch));
    await opts.maybeHeartbeat("healthy");
  });
  const heartbeatTask = (async () => {
    while (active && !opts.halt()) {
      await opts.maybeHeartbeat("healthy");
      await defaultSleep(200);
    }
  })();
  try {
    await Promise.all([accelTask, eegTask]);
  } finally {
    active = false;
    await heartbeatTask;
  }
}

async function drainIterable<T>(
  source: AsyncIterable<T>,
  halt: () => boolean,
  active: () => boolean,
  onItem: (item: T) => void | Promise<void>,
): Promise<void> {
  const iterator = source[Symbol.asyncIterator]();
  try {
    while (active() && !halt()) {
      const next = iterator.next();
      const result = await Promise.race([
        next,
        pollHalt(halt, active).then(() => ({ done: true as const, value: undefined })),
      ]);
      if (result.done) {
        return;
      }
      await onItem(result.value as T);
    }
  } finally {
    await iterator.return?.();
  }
}

async function pollHalt(halt: () => boolean, active: () => boolean): Promise<void> {
  while (active() && !halt()) {
    await defaultSleep(50);
  }
}

function normalizeEpoch(epoch: CrownEpoch): CrownEpoch {
  const record = epoch as CrownEpoch & { info?: CrownEpoch["info"]; data?: number[][] };
  return {
    data: record.data ?? [],
    info: {
      channelNames: record.info?.channelNames ?? [...CROWN_CHANNELS],
      samplingRate: record.info?.samplingRate ?? 256,
      startTime: record.info?.startTime ?? 0,
    },
  };
}

function normalizeAccel(sample: CrownAccel): CrownAccel {
  const record = sample as CrownAccel & { acceleration?: CrownAccel };
  if (record.acceleration) {
    return {
      x: record.acceleration.x,
      y: record.acceleration.y,
      z: record.acceleration.z,
    };
  }
  return { x: sample.x, y: sample.y, z: sample.z };
}

function defaultSleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, Math.max(0, ms)));
}

File: services/crown-adapter/crown_adapter/hardware.py
from __future__ import annotations

import math
import re
import time
from collections.abc import Callable
from dataclasses import dataclass, field

from intent_contracts.envelope import EventEnvelope

from crown_adapter.client import (
    CrownAccel,
    CrownAuthError,
    CrownClient,
    CrownEpoch,
    EpochQueue,
    headset_idle_detail,
    log_progress,
)
from crown_adapter.events import device_status, heartbeat, make_event
from crown_adapter.quality import (
    CROWN_CHANNELS,
    device_ms_to_ns,
    sanitize_samples,
    score_quality,
)

DEVICE_ALIAS = "crown"
LOGIN_TIMEOUT_S = 60.0
IDLE_WARN_S = 8.0
MOTION_AXES = ("x", "y", "z")


@dataclass
class CrownConfig:
    sample_rate_hz: int = 256
    channels: int = 8
    samples_per_chunk: int = 16
    heartbeat_seconds: float = 2.0
    motion_artifact_threshold: float = 0.8
    shadow_only: bool = True
    reconnect_max_seconds: float = 30.0


@dataclass
class AccelState:
    x: float = 0.0
    y: float = 0.0
    z: float = 0.0
    magnitude: float = 0.0
    has_sample: bool = False


@dataclass
class ConvertedEpoch:
    events: list[EventEnvelope] = field(default_factory=list)
    rejected: bool = False
    clock_offset_ns: int | None = None


def backoff_s(attempt: int, cap_seconds: float) -> float:
    cap = max(1.0, cap_seconds)
    return min(cap, float(2 ** max(0, attempt)))


def is_missing_credentials(message: str) -> bool:
    return any(
        token in message
        for token in (
            "NEUROSITY_EMAIL",
            "NEUROSITY_PASSWORD",
            "NEUROSITY_DEVICE_ID",
            "Crown IP",
        )
    )


def is_auth_failure(message: str) -> bool:
    if is_missing_credentials(message):
        return False
    return bool(
        re.search(
            r"password|passwd|token|secret|authorization|unauthori[sz]ed|unauthenticated|invalid.?credential",
            message,
            re.I,
        )
    )


def update_accel(target: AccelState, sample: CrownAccel) -> None:
    target.x = float(sample.x)
    target.y = float(sample.y)
    target.z = float(sample.z)
    target.magnitude = math.sqrt(target.x**2 + target.y**2 + target.z**2)
    target.has_sample = math.isfinite(target.magnitude)


def convert_epoch(
    epoch: CrownEpoch,
    accel: AccelState,
    *,
    sequence: Callable[[], int],
    packet_loss_count: int,
    chunks_emitted: int,
    motion_artifact_threshold: float,
    expected_channels: int,
    expected_samples: int,
    sample_rate_hz: float,
    shadow_only: bool,
    clock_offset_ns: int | None,
) -> ConvertedEpoch:
    samples = epoch.data if len(epoch.data) == expected_channels else []
    sample_count = len(samples[0]) if samples else 0
    if not samples or sample_count == 0:
        return ConvertedEpoch(rejected=True, clock_offset_ns=clock_offset_ns)
    channel_names = (
        epoch.channel_names
        if len(epoch.channel_names) == expected_channels
        else list(CROWN_CHANNELS)
    )
    cleaned, invalid = sanitize_samples(samples)
    source_time_ns = None
    offset = clock_offset_ns
    if epoch.start_time_ms is not None and math.isfinite(epoch.start_time_ms):
        source_time_ns = device_ms_to_ns(epoch.start_time_ms)
        if offset is None:
            offset = time.time_ns() - source_time_ns
    motion_magnitude = accel.magnitude if accel.has_sample else 0.0
    quality = score_quality(
        samples=cleaned,
        packet_loss_count=packet_loss_count,
        chunks_since_loss=chunks_emitted,
        motion_magnitude=motion_magnitude,
        motion_artifact_threshold=motion_artifact_threshold,
    )
    if invalid:
        quality.flags.append("sanitized_non_finite")
    if sample_count != expected_samples:
        quality.flags.append("unexpected_epoch_length")
    rate = epoch.sampling_rate or sample_rate_hz
    events: list[EventEnvelope] = [
        make_event(
            event_type="biosignal.chunk",
            sequence=sequence(),
            source_time_ns=source_time_ns,
            quality=quality.score,
            payload={
                "sample_rate_hz": rate,
                "channel_names": channel_names,
                "sample_count": sample_count,
                "samples": cleaned,
                "units": "microvolts",
                "filters_applied": ["crown_raw_filtered"],
                "packet_loss_count": packet_loss_count,
                "clock_confidence": 0.2 if source_time_ns is None else 1,
                "estimated_first_sample_ns": source_time_ns,
            },
        )
    ]
    if accel.has_sample:
        events.append(
            make_event(
                event_type="motion.chunk",
                modality="imu",
                sequence=sequence(),
                source_time_ns=source_time_ns,
                quality=quality.score,
                payload={
                    "sample_rate_hz": rate / max(sample_count, 1),
                    "axes": list(MOTION_AXES),
                    "sample_count": 1,
                    "samples": [[accel.x], [accel.y], [accel.z]],
                    "units": "g",
                    "magnitude": accel.magnitude,
                },
            )
        )
    events.append(
        make_event(
            event_type="data.quality",
            sequence=sequence(),
            source_time_ns=source_time_ns,
            quality=quality.score,
            payload={
                "score": quality.score,
                "components": {
                    "packet_quality": quality.packet_quality,
                    "channel_validity": quality.channel_validity,
                    "motion_penalty": quality.motion_penalty,
                },
                "flags": quality.flags,
            },
        )
    )
    if chunks_emitted > 0 and chunks_emitted % 16 == 0 and source_time_ns is not None:
        artifact = motion_magnitude > motion_artifact_threshold
        events.append(
            make_event(
                event_type="modality.feature",
                sequence=sequence(),
                source_time_ns=source_time_ns,
                quality=quality.score,
                payload={
                    "feature_name": "eeg_shadow",
                    "window_start_ns": source_time_ns,
                    "window_end_ns": source_time_ns + device_ms_to_ns(62),
                    "label": "artifact" if artifact else "ok",
                    "confidence": quality.score,
                    "candidate_scores": {
                        "ok": max(0.0, 1.0 - quality.score) if artifact else quality.score,
                        "artifact": quality.score if artifact else max(0.0, 1.0 - quality.score),
                    },
                    "model_id": "crown-shadow-v0",
                    "shadow_only": shadow_only,
                },
            )
        )
    return ConvertedEpoch(events=events, rejected=False, clock_offset_ns=offset)


class _Seq:
    def __init__(self) -> None:
        self.value = 0

    def next(self) -> int:
        current = self.value
        self.value += 1
        return current


def run_crown_hardware(
    *,
    client: CrownClient,
    send: Callable[[EventEnvelope], None],
    config: CrownConfig,
    stopped: Callable[[], bool],
    duration_s: float = 0.0,
    login_timeout_s: float = LOGIN_TIMEOUT_S,  # kept for call-site compatibility
    sleep: Callable[[float], None] = time.sleep,
    now: Callable[[], float] | None = None,
) -> None:
    clock = now or time.monotonic
    started = clock()
    _ = login_timeout_s
    deadline = started + duration_s if duration_s > 0 else 0.0
    seq = _Seq()
    last_heartbeat = float("-inf")
    attempt = 0

    def halt() -> bool:
        if stopped():
            return True
        return deadline > 0 and clock() >= deadline

    def maybe_heartbeat(status: str, last_data: float | None, dropped: int) -> None:
        nonlocal last_heartbeat
        t = clock()
        if t - last_heartbeat < config.heartbeat_seconds:
            return
        age = None if last_data is None else (t - last_data) * 1000.0
        send(heartbeat(seq.next(), t - started, age, dropped, status))
        last_heartbeat = t

    send(
        device_status(seq.next(), "degraded", "connecting to Neurosity", device_alias=DEVICE_ALIAS)
    )
    log_progress("connecting to Neurosity")
    maybe_heartbeat("degraded", None, 0)
    try:
        while not halt():
            try:
                # MindExecute connect() is blocking on the main thread.
                client.login()
                attempt = 0
                _consume_stream(
                    client=client,
                    config=config,
                    send=send,
                    seq=seq,
                    halt=halt,
                    maybe_heartbeat=maybe_heartbeat,
                    sleep=sleep,
                    clock=clock,
                )
                if halt():
                    break
                send(
                    device_status(
                        seq.next(),
                        "degraded",
                        "crown stream ended; reconnecting",
                        device_alias=DEVICE_ALIAS,
                    )
                )
            except CrownAuthError as exc:
                detail = str(exc)
                log_progress(detail)
                if is_missing_credentials(detail):
                    send(
                        device_status(
                            seq.next(),
                            "offline",
                            "missing Crown OSC IP",
                            device_alias=DEVICE_ALIAS,
                        )
                    )
                    print(_safe_error(detail), flush=True)
                    return
                if "timed out" in detail.lower():
                    send(
                        device_status(
                            seq.next(),
                            "degraded",
                            "Crown connect timed out",
                            {"reconnect_attempt": attempt + 1},
                            DEVICE_ALIAS,
                        )
                    )
                elif is_auth_failure(detail) or "authentication failed" in detail.lower():
                    send(
                        device_status(
                            seq.next(),
                            "degraded",
                            "neurosity authentication failed; reconnecting",
                            {"reconnect_attempt": attempt + 1},
                            DEVICE_ALIAS,
                        )
                    )
                else:
                    send(
                        device_status(
                            seq.next(),
                            "degraded",
                            detail,
                            {"reconnect_attempt": attempt + 1},
                            DEVICE_ALIAS,
                        )
                    )
            except Exception as exc:
                log_progress(f"crown disconnected ({type(exc).__name__})")
                send(
                    device_status(
                        seq.next(),
                        "degraded",
                        "crown disconnected; reconnecting",
                        {"reconnect_attempt": attempt + 1},
                        DEVICE_ALIAS,
                    )
                )
                _ = exc
            if halt():
                break
            wait_s = backoff_s(attempt, config.reconnect_max_seconds)
            attempt += 1
            log_progress(f"retry in {wait_s:.0f}s")
            try:
                client.stop()
            except Exception:
                pass
            wait_end = clock() + wait_s
            while not halt() and clock() < wait_end:
                maybe_heartbeat("degraded", None, 0)
                sleep(min(0.2, wait_end - clock()))
    finally:
        send(device_status(seq.next(), "offline", "adapter stopping", device_alias=DEVICE_ALIAS))
        try:
            client.stop()
        except Exception:
            pass


def _consume_stream(
    *,
    client: CrownClient,
    config: CrownConfig,
    send: Callable[[EventEnvelope], None],
    seq: _Seq,
    halt: Callable[[], bool],
    maybe_heartbeat: Callable[[str, float | None, int], None],
    sleep: Callable[[float], None],
    clock: Callable[[], float],
) -> None:
    pending = EpochQueue()
    accel = AccelState()
    last_start = float("-inf")
    chunks_emitted = 0
    packet_loss_count = 0
    clock_offset_ns: int | None = None
    last_data: float | None = None
    waiting_since = clock()
    idle_notified = False
    send(
        device_status(
            seq.next(),
            "degraded",
            "logged in; waiting for EEG samples",
            device_alias=DEVICE_ALIAS,
        )
    )
    client.start(pending.push_epoch, pending.push_accel)
    poll_logged = False
    try:
        while not halt():
            update_accel(accel, pending.accel())
            epochs = pending.drain_epochs()
            if not epochs:
                try:
                    polled = client.poll_epoch()
                except Exception:
                    polled = None
                if polled is not None:
                    if not poll_logged:
                        log_progress("receiving EEG via Firebase poll (live stream idle)")
                        poll_logged = True
                    epochs = [polled]
            if (
                not epochs
                and chunks_emitted == 0
                and not idle_notified
                and clock() - waiting_since >= IDLE_WARN_S
            ):
                idle_notified = True
                try:
                    status = client.headset_status()
                except Exception:
                    status = {}
                if accel.has_sample:
                    detail = (
                        "Crown is online but not publishing EEG; "
                        "put it on your head until the sensors make contact"
                    )
                else:
                    detail = headset_idle_detail(status if isinstance(status, dict) else {})
                log_progress(detail)
                send(
                    device_status(
                        seq.next(),
                        "degraded",
                        detail,
                        device_alias=DEVICE_ALIAS,
                    )
                )
            for epoch in epochs:
                start = epoch.start_time_ms
                if start is not None and math.isfinite(start) and start <= last_start:
                    continue
                if start is not None and math.isfinite(start):
                    last_start = start
                converted = convert_epoch(
                    epoch,
                    accel,
                    sequence=seq.next,
                    packet_loss_count=packet_loss_count + pending.dropped,
                    chunks_emitted=chunks_emitted,
                    motion_artifact_threshold=config.motion_artifact_threshold,
                    expected_channels=config.channels,
                    expected_samples=config.samples_per_chunk,
                    sample_rate_hz=config.sample_rate_hz,
                    shadow_only=config.shadow_only,
                    clock_offset_ns=clock_offset_ns,
                )
                if converted.rejected:
                    packet_loss_count += 1
                    continue
                clock_offset_ns = converted.clock_offset_ns
                for event in converted.events:
                    send(event)
                if chunks_emitted == 0:
                    send(
                        device_status(
                            seq.next(), "healthy", "crown stream started", device_alias=DEVICE_ALIAS
                        )
                    )
                    log_progress("EEG stream started (8 ch, 256 Hz)")
                chunks_emitted += 1
                last_data = clock()
            maybe_heartbeat(
                "healthy" if chunks_emitted else "degraded",
                last_data,
                packet_loss_count + pending.dropped,
            )
            if chunks_emitted > 0 and not epochs:
                # Live SDK pushes from a background thread; sleep between polls.
                # Fake clients dump epochs during start(); one extra drain pass is enough.
                if pending.drain_epochs():
                    continue
                if not hasattr(client, "epochs"):
                    sleep(0.05)
                else:
                    return
            elif not epochs:
                sleep(0.05)
            elif not hasattr(client, "epochs"):
                sleep(0.05)
    finally:
        try:
            client.stop()
        except Exception:
            pass


def _safe_error(raw: str) -> str:
    if is_missing_credentials(raw):
        return (
            '{"level":"error","service":"crown-adapter",'
            '"msg":"missing Crown OSC IP; pass --ip or set devices.crown.ip_address '
            '(same as MindExecute --ip)"}'
        )
    if is_auth_failure(raw) or "authentication failed" in raw.lower():
        return (
            '{"level":"error","service":"crown-adapter",'
            '"msg":"Neurosity authentication failed; check .env.local (values are not logged)"}'
        )
    return raw
